code/network/lib/Loot/Network/ZMQ/Adapter.hs

```diff
@@ -0,0 +1,182 @@
+-- is actually any data to read from it.
+threadWaitReadSTMLong :: Fd -> IO (STM (), IO ())
+threadWaitReadSTMLong fd = do
+    m <- Sync.newTVarIO False
+    t <- Sync.forkIO $ forever $ do
+        G.threadWaitRead fd
+        Sync.atomically $ Sync.writeTVar m True
+    let waitAction = do b <- Sync.readTVar m
+                        if b then Sync.writeTVar m False else retry
+    let killAction = Sync.killThread t
+    return (waitAction, killAction)
+
+-- | 'threadWaitReadSTMLong' adapted for sockets.
+socketWaitReadSTMLong :: (MonadIO m) => Z.Socket t -> m (STM (), m ())
+socketWaitReadSTMLong s = liftIO $ do
+    fd <- Z.fileDescriptor s
+    (stm, destroy) <- threadWaitReadSTMLong fd
+    pure (stm, liftIO destroy)
+
+-- | Checks if data can be received from the socket. Use @whileM
+-- canReceive process@ pattern after the STM action on the needed
+-- socket.
+canReceive :: Z.Socket t -> IO Bool
+canReceive sock = elem Z.In <$> Z.events sock
+
+
+{-
+http://api.zeromq.org/master:zmq-getsockopt#toc30
+
+The combination of a file descriptor returned by the ZMQ_FD option
+being ready for reading but no actual events returned by a subsequent
+retrieval of the ZMQ_EVENTS option is valid; applications should
+simply ignore this case and restart their polling operation/event
+loop.
+
+Issues:
+https://github.com/zeromq/libzmq/issues/1434
+https://github.com/zeromq/czmq/issues/1487
+
+Mans:
+https://funcptr.net/2013/04/20/embedding-zeromq-in-the-libev-event-loop/
+https://funcptr.net/2012/09/10/zeromq---edge-triggered-notification/
+
+-}
+
+----------------------------------------------------------------------------
+-- Tests TODO move to benches (?)
+----------------------------------------------------------------------------
+
+_testWait :: Z.Socket Z.Router -> IO ()
+_testWait sock = do
+    putTextLn "testWait: starting"
+    fd <- Z.fileDescriptor sock
+    let action = do
+            putTextLn "testWait: waiting"
+            G.threadWaitRead fd
+            putTextLn "testWait: waiting exited"
+            whileM (canReceive sock) $ do
+              putTextLn "testWait: receiving"
+              d <- Z.receiveMulti sock
+              putTextLn $ "testWait: " <> show d
+    forever action
+
+_testWaitSTM :: Z.Socket Z.Router -> IO ()
+_testWaitSTM sock = do
+    putTextLn "testWaitStm: waiting"
+    fd <- Z.fileDescriptor sock
+    let action = do
+            putTextLn "testWaitStm: waiting"
+            (waitAct,afterWaitAct) <- G.threadWaitReadSTM fd
+            atomically waitAct `finally` afterWaitAct
+            putTextLn "testWaitStm: waiting exited"
+            whileM (canReceive sock) $ do
+              putTextLn "testWaitStm: receiving"
+              d <- Z.receiveMulti sock
+              putTextLn $ "testWaitStm: " <> show d
+    forever action
+
+
+type Msg = NonEmpty ByteString
+
+data WaitRes  = ReadSock | ReadQueue Msg deriving Show
+
+pollBoth :: Z.Socket Z.Dealer -> TQueue Msg -> TQueue Msg -> IO ()
```

Comment:
From the type, it is hard to understand which queue is input one.
And therefore, it is easy to mess up.
You can once make a `ZMQConfig` and save yourself (or someone other) a lot of debugging time.

Reply:
This is an internal function for testing.